https://issues.redhat.com/browse/ACM-25316 Add TNF AI docs (#8425)
open-cluster-management/rhacm-docs: clusters/assisted_installer/ai_install_tnf.adoc

[#install-tnf-cluster-ai]

> = Installing a two nodes with fencing cluster with {ai} (Technology Preview)

Reviewer: can we put (TNF) here because reading this out loud, and how it will be translated or read with a screen reader, it is confusing:

"You can now install a two nodes with fencing cluster" appears at first like the "a" should not be there, for instance.

Author: Added

Author: OCP Docs uses `two-node with fencing cluster`. Do we want to follow them or go our own way with TNF?

A two nodes with fencing (TNF) cluster is an {ocp} cluster that has two control plane nodes. Use TNF clusters to ensure redundancy at the edge and limit hardware costs.

.Prerequisites

- You are using the `baremetal` or `none` platform.
- You have access to the baseboard management controller (BMC) of the platform nodes.

> - Your {ocp-short} managed cluster version is 4.20 or later.

Reviewer: We need to remove this: (or later) 

bc there is only 4.20. 

You can say 4.20, but the user could be stopped here and asking where is "later" releases and how do they get to them. We just will have to edit it next release.

Author: I'll remove it, I'll make a note to add it back for 2.16

.Procedure

Complete the following steps to install a TNF cluster by using {ai}:

. Enable the TNF cluster installation feature with {ai} by setting an environment variable on the `assisted-service` pod.

.. Create the following `ConfigMap` resource:
+
[source,yaml]
----
apiVersion: v1
data:
  TNF_CLUSTERS_SUPPORT: "true"
kind: ConfigMap
metadata:
  name: assisted-service-custom
  namespace: multicluster-engine
----

.. Reference the `ConfigMap` resource in an annotation on the `AgentServiceConfig` resource. Run the following command:
+
[source,bash]
----
oc annotate agentserviceconfig agent unsupported.agent-install.openshift.io/assisted-service-configmap=assisted-service-custom
----

. Configure the TNF cluster object on the hub cluster by creating the required resources.

.. Create a namespace. Use the following YAML sample:
+
[source,yaml]
----
apiVersion: v1
kind: Namespace
metadata:
  name: tnf-cluster
----

> .. Create a pull secret. Use the following YAML sample. Replace values where needed:

Reviewer: Replace values where needed: (we should be specific). I have seen this entering our doc recently, but we need to tell them what to replace.

For instance:

Replace `<your-base64-pull-secret>` with the name of your pull secret.

Author: I'll update it across the file

+
[source,yaml]
----
apiVersion: v1
kind: Secret
type: kubernetes.io/dockerconfigjson
metadata:
  name: pull-secret
  namespace: tnf-cluster
data:
  .dockerconfigjson: <your-base64-pull-secret>
----

> .. Create a `AgentClusterInstall` resource. Use the following YAML sample. Replace values where needed:

Reviewer: Same here

+
[source,yaml]
----
apiVersion: extensions.hive.openshift.io/v1beta1
kind: AgentClusterInstall
metadata:
  name: tnf-cluster
  namespace: tnf-cluster
spec:
  platformType: BareMetal
  apiVIPs: [192.168.122.10]
  ingressVIPs: [192.168.122.20]
  clusterDeploymentRef:
    name: tnf-cluster
  imageSetRef:
    name: img4.20.0-rc.3-x86-64-appsub
  networking:
    clusterNetwork:
    - cidr: 10.128.0.0/14
      hostPrefix: 23
    machineNetwork:
    - cidr: 192.168.122.0/24
    serviceNetwork:
    - 172.30.0.0/16
  provisionRequirements:
    controlPlaneAgents: 2
  sshPublicKey: <your-public-ssh-key>
----

.. Create a `ClusterDeployment` resource. Use the following YAML sample. Replace values where needed:

> +

Reviewer: Same here with replace.

Author: Updated

  name: tnf-cluster
  namespace: tnf-cluster
spec:
  baseDomain: <yourdomain.com>
  clusterInstallRef:
    group: extensions.hive.openshift.io
    kind: AgentClusterInstall
    name: tnf-cluster
    version: v1beta1
  clusterName: tnf-cluster
  controlPlaneConfig:
    servingCertificates: {}
  platform:
    agentBareMetal:
      agentSelector:
        cluster: tnf
  pullSecretRef:
    name: pull-secret
----

.. Create an `InfraEnv` resource. Use the following YAML sample. Replace values where needed:
+
[source,yaml]
----
apiVersion: agent-install.openshift.io/v1beta1
kind: InfraEnv
metadata:
  name: tnf-cluster
  namespace: tnf-cluster
spec:
  clusterRef:
    name: tnf-cluster
    namespace: tnf-cluster
  pullSecretRef:
    name: pull-secret
  sshAuthorizedKey: <your-public-ssh-key>
----

. *Optional:* Create a fencing credentials secret for each node. Use the following YAML sample. Replace values where needed:

+
[source,yaml]
----
apiVersion: v1
stringData:
  address: https://bmc.example.com
  certificateVerification: Disabled
  username: <username>
  password: <password>
kind: Secret
metadata:
  name: node-0-fencing-credentials
  namespace: tnf-cluster
type: Opaque
----

> - *Optional:* The `certificateVerification` field default value is `Enabled`.

Author: Red Hat style guide tells us to do call outs like this. Do we want to follow that?

Reviewer: you can but you can also make it part of the step. Not sure why we have optional here?

Author: I'll move it to the step

+
*Note:* Fencing credentials for an Agent are stored as a Kubernetes secret in the Agent namespace that the Agent references in the `agent.spec.fencingCredentialsSecretRef` field.

. Set the fencing credentials to approve two Agents. Use one of the following methods:

> - Manually create the secret and add an annotation on the Agent `BareMetalHost` resource that references the secret. Use the following annotation key:

Author: Not sure if a bullet list is the best way to indicate different options, let me know what you think @swopebe 

Mostly concerned since each bullet point is still a step/part of the procedure

Reviewer: I think the bulleted list works. What is better I think is to provide the reasons to pick one or the other. I am not sure why there are two options at this point, is what I am saying. Something like

. Set the ____ . Use one of the two following steps, depending on __:

.. If ___, manually create...

.. If ___, set an annotation...

You can also do that with the bullets, we may just have to change it later for conversion to DITA-- not sure. We may not be able to create bullets under a step. Don't remember exactly from my days in DITA.

Author: It's listed in order of priority, I can add that info to the step

+
[source,bash]

> ----

Reviewer: watch format here

Author: Looks OK on my end, what exactly is off?

Reviewer: these should be grayed out.

Reviewer: I just pressed enter and it worked.

Author: The code is part of the bullet list, so yellow is correct for these.

The format currently with the space is off:

<img width="2108" height="938" alt="image" src="https://github.com/user-attachments/assets/8c3aff10-d139-4ce9-a5f6-8c6cd897d8cb" />

bmac.agent-install.openshift.io/fencing-credentials-secret-name
----
+
*Note:* The {ai} sets the Agent `fencingCredentialsSecretRef` field to the value of the annotation.

- Set an annotation on the Agent `BareMetalHost` resource to create the secret from the `BareMetalHost` BMC configuration. Use the following annotation key:

> +

Reviewer: format

Reviewer: did the same here.

[#install-tnf-cluster-ai]
= Installing a two nodes with fencing cluster with {ai} (Technology Preview)

A two nodes with fencing (TNF) cluster is an {ocp} cluster that has two control plane nodes. Use TNF clusters to ensure redundancy at the edge and limit hardware costs.

.Prerequisites

- You are using the `baremetal` or `none` platform.

> - You have 2 control plane nodes configured.

Reviewer 2: I don't think this is a prereq since these hosts will be added using the infraenv we create in this procedure, right?

Author: I went ahead and removed it

- You have access to the baseboard management controller (BMC) of the platform nodes.

> - Your {ocp-short} version is 4.20 or later.

Reviewer 2: This should probably be more clear that it means that the cluster we're _installing_ needs to be 4.20 or later (as opposed to the version MCE is running on)

Reviewer 3: Yes all of the prerequisites are for the spoke cluster

Author: Updated!

.Procedure

Complete the following steps to install a TNF cluster by using {ai}:

. Enable the TNF cluster installation feature with {ai} by setting an environment variable on the `assisted-service` pod.

.. Create the following `ConfigMap` resource:
+
[source,yaml]
----
apiVersion: v1
data:
  TNF_CLUSTERS_SUPPORT: "true"
kind: ConfigMap
metadata:
  name: assisted-service-custom
  namespace: multicluster-engine
----

.. Reference the `ConfigMap` resource in an annotation on the `AgentServiceConfig` resource. Run the following command:
+
[source,bash]
----

> oc annotate agentserviceconfig agent unsupported.agent-install.openshift.io/assisted-service-configmap=assisted-service-custom

Reviewer 2: Is this here because it's a tech preview feature? Maybe we should mention here that's why this is necessary?

Reviewer 3: It's mentioned at the top that it's tech preview, maybe we should also mention it here and say that's why it's disabled by default?

Author: If tech preview is mentioned at the beginning, it applies to the entire section. We wouldn't repeat it again in the procedure as per our current doc guidelines.

If it's too confusing for customers we can maybe revisit this or add a note if necessary.

Reviewer: If you mention here that this is tech preview, you suggest to the user that the rest of the procedure is not tech preview, because you highlight one step as tech preview. That is why we put it in the title. 

We should also put it in the short description:

**Technology Preview:** short description....

Author: I added it to the description

----

. Configure the TNF cluster object on the hub cluster by creating the required resources.

.. Create a namespace. Use the following YAML sample:
+
[source,yaml]
----
apiVersion: v1
kind: Namespace
metadata:
  name: tnf-cluster
----

.. Create a pull secret. Use the following YAML sample. Replace values where needed:
+
[source,yaml]
----
apiVersion: v1
kind: Secret
type: kubernetes.io/dockerconfigjson
metadata:
  name: pull-secret
  namespace: tnf-cluster
data:
  .dockerconfigjson: <your-base64-pull-secret>
----

.. Create a `AgentClusterInstall` resource. Use the following YAML sample. Replace values where needed:
+
[source,yaml]
----
apiVersion: extensions.hive.openshift.io/v1beta1
kind: AgentClusterInstall
metadata:
  name: tnf-cluster
  namespace: tnf-cluster
spec:
  platformType: BareMetal
  apiVIPs: [192.168.122.10]
  ingressVIPs: [192.168.122.20]
  clusterDeploymentRef:
    name: tnf-cluster
  imageSetRef:
    name: img4.20.0-rc.3-x86-64-appsub
  networking:
    clusterNetwork:
    - cidr: 10.128.0.0/14
      hostPrefix: 23
    machineNetwork:
    - cidr: 192.168.122.0/24
    serviceNetwork:
    - 172.30.0.0/16
  provisionRequirements:

>     controlPlaneAgents: 2

Reviewer 2: https://docs.redhat.com/en/documentation/red_hat_advanced_cluster_management_for_kubernetes/2.14/html/clusters/cluster_mce_overview#create-agentclusterinstall-cr this lists the options for this field as 1 or 3. Maybe we should just roll TNF into the existing doc? The only differences are this field, and the BMC secret, right?

Reviewer 3: The differences are this field, the secrets at the end of this doc, and the prerequisites

Author: I'm OK with moving the entire doc set if it makes sense, right now I don't see a strong preference for either options.

@carbonin @giladravid16 What's the consensus?

Reviewer: Keep in mind our docs have to be chunked up into smaller files going forward, with one or MAYBE two concepts per file. This is for JTBD and for conversion.

Author: I'll leave this as is for now

  sshPublicKey: <your-public-ssh-key>
----

.. Create a `ClusterDeployment` resource. Use the following YAML sample. Replace values where needed:
+
[source,yaml]
----
apiVersion: hive.openshift.io/v1
kind: ClusterDeployment
metadata:
  name: tnf-cluster
  namespace: tnf-cluster
spec:
  baseDomain: <yourdomain.com>
  clusterInstallRef:
    group: extensions.hive.openshift.io
    kind: AgentClusterInstall
    name: tnf-cluster
    version: v1beta1
  clusterName: tnf-cluster
  controlPlaneConfig:
    servingCertificates: {}
  platform:
    agentBareMetal:
      agentSelector:
        cluster: tnf
  pullSecretRef:
    name: pull-secret
----

> .. Create a `ClusterDeployment` resource. Use the following YAML sample. Replace values where needed:

Reviewer 2: This should be `InfraEnv`

Author: Fixed!

+
[source,yaml]
----
apiVersion: agent-install.openshift.io/v1beta1
kind: InfraEnv
metadata:
  name: tnf-cluster
  namespace: tnf-cluster
spec:
  clusterRef:
    name: tnf-cluster
    namespace: tnf-cluster
  pullSecretRef:
    name: pull-secret
  sshAuthorizedKey: <your-public-ssh-key>
----

> . Create a fencing credentials secret. Use the following YAML sample. Replace values where needed:

Reviewer 3: This should be done for each node, but this step should be marked as optional because of the methods mentioned bellow (one of them doesn't require users to create these secrets)

Author: Updated!

+
[source,yaml]
----
apiVersion: v1
stringData:
  address: https://bmc.example.com
  certificateVerification: Disabled
  username: <username>
  password: <password>
kind: Secret
metadata:
  name: node-0-fencing-credentials
  namespace: tnf-cluster
type: Opaque
----
- *Optional:* The `certificateVerification` field default value is `Enabled`.
+
*Note:* Fencing credentials for an Agent are stored as a Kubernetes secret in the Agent namespace that the Agent references in the `agent.spec.fencingCredentialsSecretRef` field.

. Set the fencing credentials to approve two Agents. Use one of the following methods:

- Manually create the secret and add an annotation on the Agent `BareMetalHost` resource that references the secret. Use the following annotation key:
+
[source,bash]
----
bmac.agent-install.openshift.io/fencing-credentials-secret-name
----
+
*Note:* The {ai} sets the Agent `fencingCredentialsSecretRef` field to the value of the annotation.

> - Set an annotation on the Agent `BareMetalHost` resource to create the secret from the `BareMetalHost` BMC configuration. Then, set the Agent `fencingCredentialsSecretRef` to the secret name, which is `<your-agent-name>-fencing-credentials`. Use the following annotation key:

Reviewer 3: In this method the users don't need to update the Agent themselves, Assisted does it for them

Author: Removed

+
[source,bash]
----
bmac.agent-install.openshift.io/create-fencing-credentials-secret
----
+
*Note:* You can use any value.

> - Manually create the secret. Then, manually set the Agent zfencingCredentialsSecretRef` field to the secret name.

Reviewer 2: ```suggestion
- Manually create the secret. Then, manually set the Agent `fencingCredentialsSecretRef` field to the secret name.
```

I think this is a typo?

Author: It is, good catch!

      agentSelector:
        cluster: tnf
  pullSecretRef:
    name: pull-secret
----

.. Create an `InfraEnv` resource. Use the following YAML sample. Replace `<your-public-ssh-key>` with your public SSH key:
+
[source,yaml]
----
apiVersion: agent-install.openshift.io/v1beta1
kind: InfraEnv
metadata:
  name: tnf-cluster
  namespace: tnf-cluster
spec:
  clusterRef:
    name: tnf-cluster
    namespace: tnf-cluster
  pullSecretRef:
    name: pull-secret
  sshAuthorizedKey: <your-public-ssh-key>
----

. *Optional:* Create a fencing credentials secret for each node. Use the following YAML sample. Replace `<username>` and `<password>` with your username and password. The `certificateVerification` field default value is `Enabled`:

+
[source,yaml]
----
apiVersion: v1
stringData:
  address: https://bmc.example.com
  certificateVerification: Disabled
  username: <username>
  password: <password>
kind: Secret
metadata:
  name: node-0-fencing-credentials
  namespace: tnf-cluster
type: Opaque
----
+
*Note:* Fencing credentials for an Agent are stored as a Kubernetes secret in the Agent namespace that the Agent references in the `agent.spec.fencingCredentialsSecretRef` field.

> . Set the fencing credentials to approve two Agents. Use one of the following methods. The methods are listed in descending order of priority:

Reviewer 2: Do the credentials have anything to do with approving agents?
This first sentence is a bit strange to me. Maybe just
```suggestion
. Use one of the following methods to set Agent fencing credentials. The methods are listed in descending order of priority:
```

> - Manually create the secret and add an annotation on the Agent `BareMetalHost` resource that references the secret. Use the following annotation key:

Reviewer 2: `Agent` and `BareMetalHost` are two different CRs that represent very similar concepts. Maybe just drop "Agent" here for clarity?
```suggestion
- Manually create the secret and add an annotation on the `BareMetalHost` resource that references the secret. Use the following annotation key:
```

+
[source,bash]
----
bmac.agent-install.openshift.io/fencing-credentials-secret-name
----
+
*Note:* The {ai} sets the Agent `fencingCredentialsSecretRef` field to the value of the annotation.

> - Set an annotation on the Agent `BareMetalHost` resource to create the secret from the `BareMetalHost` BMC configuration. Use the following annotation key:

Reviewer 2: Can this use any value? @giladravid16 ?
